src/main/java/com/google/cloud/spanner/pgadapter/statements/IntermediateStatement.java

```diff
@@ -264,6 +299,78 @@ public void execute() {
     }
   }
 
+  /**
+   * Executes all the statements in this {@link IntermediateStatement}. This method tries to batch
+   * statements together when that is possible (i.e. multiple DML statements in a row), and executes
+   * other statements individually.
+   *
+   * @throws SQLException If a database error occurs
+   */
+  private void executeBatch() throws SQLException {
+    Preconditions.checkState(this.statements.size() > 1);
+    int index = 0;
+    while (index < getStatementCount()) {
+      boolean canUseBatch = false;
+      if (index < (getStatementCount() - 1)) {
+        StatementType statementType = getStatementType(index);
+        StatementType nextStatementType = getStatementType(index + 1);
+        canUseBatch = canBeBatchedTogether(statementType, nextStatementType);
+      }
+      if (canUseBatch) {
+        index += executeStatementsInBatch(index);
+      } else {
+        this.statement.execute(statements.get(index));
+        updateResultCount(index);
+        index++;
+      }
+    }
+  }
+
+  /**
+   * Executes the statements from fromIndex in a DML/DDL batch. The batch will consist of all
+   * statements from fromIndex till the first statement that is of a different type than the
+   * statement at fromIndex. That is; If the first statement is a DML statement, the batch will
+   * contain all statements that follow until it encounters a statement that is not a DML statement.
+   * The same also applies to DDL statements. Query statements and other statements can not be
+   * batched.
+   *
+   * @param fromIndex The index of the statements array where the batch should start
+   * @return The number of statements included in the batch.
+   */
+  private int executeStatementsInBatch(int fromIndex) throws SQLException {
+    Preconditions.checkArgument(fromIndex < getStatementCount() - 1);
+    Preconditions.checkArgument(
+        canBeBatchedTogether(getStatementType(fromIndex), getStatementType(fromIndex + 1)));
+    StatementType batchType = getResultTypes().get(fromIndex);
+    // Add the first statement to the batch, as we know that this statement is compatible with this
+    // batch.
+    this.statement.addBatch(statements.get(fromIndex));
+    int index = fromIndex + 1;
+    while (index < getStatementCount()) {
+      StatementType statementType = getStatementType(index);
+      if (canBeBatchedTogether(batchType, statementType)) {
+        statement.addBatch(statements.get(index));
+        index++;
+      } else {
+        // End the batch here, as the statement type on this index can not be batched together with
+        // the other statements in the batch.
+        break;
+      }
+    }
+    int[] updateCounts = this.statement.executeBatch();
```

Comment:
No, that is only true for DDL batches. DML batches return the actual update counts:

From the `Connection.runBatch()` documentation:
```
Returns:
the update counts in case of a DML batch. Returns an array containing 1 for each successful statement and 0 for each failed statement or statement that was not executed in case of a DDL batch.
```